### PR TITLE
SE: Improve XOROperation

### PR DESCRIPTION
**Pull request by Tim-Pohlmann**
Improvement for #7246

### CODE REVIEW

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/Binary.Arithmetic.cs [hunk 1]
```diff
@@ -257,13 +271,23 @@ private static NumberConstraint AccountForZero(NumberConstraint constraint)
...
+        else if ((left.IsPositive || right.IsPositive) && left.Max.HasValue && right.Max.HasValue)
         {
-            return -1;
+            return PositiveMagnitude(BigInteger.Max(left.Max.Value, right.Max.Value));
+        }
+        else if ((left.IsNegative || right.IsNegative) && left.Min.HasValue && right.Min.HasValue)
+        {
+            return PositiveMagnitude(BigInteger.Max(BigInteger.Abs(left.Min.Value), BigInteger.Abs(right.Min.Value)));
+        }
+        else if (left.Min.HasValue && left.Max.HasValue && right.Min.HasValue && right.Max.HasValue)
```
Comment (antonioaversa) [inline analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/Binary.Arithmetic.cs:286]: This branch is not covered by test cases, and that makes the quality gate on SC fail.

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/Binary.Arithmetic.cs [hunk 2]
```diff
@@ -249,6 +249,20 @@ private static NumberConstraint AccountForZero(NumberConstraint constraint)
                 return 0;
             }
         }
+        else if ((left.IsPositive || right.IsNegative) && left.Max.HasValue && right.Min.HasValue)
```
Comment (antonioaversa) [inline analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/Binary.Arithmetic.cs:252]: The code smell on this method suggests that the complexity of it should be reduced to <= 15, and I agree that the function is now difficult to follow due to the many branches and the logic operators, sometimes nested.

I think giving semantics to the following conditions, via local functions, would help:
- `(left.IsNegative || right.IsPositive) && left.Min.HasValue && right.Max.HasValue`
- `left.Min.HasValue && left.Max.HasValue && right.Min.HasValue && right.Max.HasValue`

Extracting nested if blocks on the first two branches would also help (in the first version of this code I had `CanOverlap` and a more concise version of this, which didn't have nested conditional statements on the same variables).

Local functions are not included in the calculation of the cognitive complexity of the parent method.

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/Binary.Arithmetic.cs [hunk 3]
```diff
@@ -257,13 +271,23 @@ private static NumberConstraint AccountForZero(NumberConstraint constraint)
...
-            return -1;
+            return PositiveMagnitude(BigInteger.Max(left.Max.Value, right.Max.Value));
+        }
+        else if ((left.IsNegative || right.IsNegative) && left.Min.HasValue && right.Min.HasValue)
+        {
+            return PositiveMagnitude(BigInteger.Max(BigInteger.Abs(left.Min.Value), BigInteger.Abs(right.Min.Value)));
+        }
+        else if (left.Min.HasValue && left.Max.HasValue && right.Min.HasValue && right.Max.HasValue)
+        {
+            return PositiveMagnitude(BigInteger.Max(
```
Comment (antonioaversa) [inline analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/Binary.Arithmetic.cs:288]: Does the multi-level max have semantics here, or it's just lack of a Max of more than 2 values?
If it has semantics: a comment would help understanding what that is about.
If it hasn't, what about a private static method `Max` with params args, or the max of an array?

**Review by antonioaversa — CHANGES_REQUESTED** [review]
Coverage and code smell about cognitive complexity to be handled.

File: analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Binary.cs [hunk 4]
```diff
@@ -1082,8 +1082,9 @@ public void Binary_BitXor_SingleValue(int left, int right, int expected)
     [DataRow("i >=  4 && i <=  6 && j >=  6", 0, null)]
     [DataRow("i >=  4 && i <=  6 && j >=  6 && j <= 8", 0, 15)]   // exact range: 0, 14
     [DataRow("i >=  4 && i <=  5 && j >=  6 && j <= 8", 1, 15)]   // exact range: 2, 13
-    [DataRow("i >= -3 && i <= -1 && j >= -4 && j <=-2", 0, null)] // exact range: 0, 3
-    [DataRow("i >= -4 && i <= -3 && j >= -2 && j <=-1", 1, null)] // exact range: 2, 3
+    [DataRow("i >= -3 && i <= -1 && j >= -4 && j <=-2", 0, 7)]    // exact range: 0, 3
+    [DataRow("i >= -4 && i <= -3 && j >= -2 && j <=-1", 1, 7)]    // exact range: 2, 3
+    [DataRow("i >= -4 && i <=  6 && j >= -3 && j <= 8", -8, 15)]
```
Comment (antonioaversa) [inline analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Binary.cs:1087]: The exact range for the new interval is $`[-12, 14]`$.
The range found is $`[-8, 15]`$, which doesn't contain the exact range on the negative side.
If we take -12, for example: $`-4 \oplus 8 = -12 \notin [-8, 15]`$

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/Binary.Arithmetic.cs [hunk 5]
```diff
@@ -217,60 +217,76 @@ private static NumberConstraint AccountForZero(NumberConstraint constraint)
...
+        else if (left.Min.HasValue && left.Max.HasValue && right.Min.HasValue && right.Max.HasValue)
+        {
+            return NegativeMagnitude(-Max(BigInteger.Abs(left.Min.Value), left.Max.Value, BigInteger.Abs(right.Min.Value), right.Max.Value));
+        }
+        else
+        {
+            return null;
+        }
+
+        static BigInteger? SameSign(NumberConstraint range1, NumberConstraint range2)
```
Comment (antonioaversa) [inline analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/Binary.Arithmetic.cs:245]: I like the comment and the function itself

**Review by antonioaversa — CHANGES_REQUESTED** [review]
Very close to the correct implementation! 
There is only one case which, as far as I can see, is still not correct.

**Comment by antonioaversa** [comment]
Also, the `Sonar.NET` check is failing, for some reason.

File: analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Binary.cs (hunk 4 above)
Comment (Tim-Pohlmann) [reply analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Binary.cs]: Very good catch. Nasty edge cases. I added a comment for further explanations.

Comment (antonioaversa) [reply analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Binary.cs]: Very good!

**Review by antonioaversa — APPROVED** [review]
LGTM, great!

**Review by antonioaversa — APPROVED** [review]
LGTM, great!
